### PR TITLE
JDK-6251738: Want a top-level summary page that itemizes all spec documents referenced from javadocs (OEM spec)

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review the code and tests to add support for a new `@spec url title` tag to javadoc.  Note, this does not include any changes to API doc comments to use the new tag in JDK API documentation; that will come later.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-6251738](https://bugs.openjdk.org/browse/JDK-6251738): Want a top-level summary page that itemizes all spec documents referenced from javadocs (OEM spec)
 * [JDK-8264740](https://bugs.openjdk.org/browse/JDK-8264740): Want a top-level summary page that itemizes all spec documents referenced from javadocs (OEM spec) (**CSR**)

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8439/head:pull/8439` \
`$ git checkout pull/8439`

Update a local copy of the PR: \
`$ git checkout pull/8439` \
`$ git pull https://git.openjdk.org/jdk pull/8439/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8439`

View PR using the GUI difftool: \
`$ git pr show -t 8439`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8439.diff">https://git.openjdk.org/jdk/pull/8439.diff</a>

</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ExternalSpecsWriter.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+import jdk.javadoc.internal.doclets.toolkit.util.DocPaths;
+import jdk.javadoc.internal.doclets.toolkit.util.IndexItem;
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Generates the file with the summary of all the references to external specifications.
+ *
+ *  <p><b>This is NOT part of any supported API.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ExternalSpecsWriter.java:68]: We abolished such notes in JDK-8284362.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDoclet.java [hunk 2]
```diff
@@ -30,7 +30,14 @@
 import java.nio.file.Files;
 import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
-import java.util.*;
+import java.util.ArrayList;
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDoclet.java:33]: This does not seem necessary for this PR and looks like an artifact of editing in an IDE.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import jdk.javadoc.internal.doclets.toolkit.Content;
+import jdk.javadoc.internal.doclets.toolkit.util.CommentHelper;
+import jdk.javadoc.internal.doclets.toolkit.util.DocFinder;
+import jdk.javadoc.internal.doclets.toolkit.util.DocFinder.Input;
+import jdk.javadoc.internal.doclets.toolkit.util.Utils;
+
+/**
+ * A taglet that represents the {@code @spec} tag.
+ *
+ *  <p><b>This is NOT part of any supported API.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:44]: See my comment on such notes above.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ *  If you write code that depends on this, you do so at your own risk.
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class SpecTaglet extends BaseTaglet implements InheritableTaglet {
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:49]: Ooh! This taglet is inheritable. Is it the right thing to do? If the inheritance is not needed down the hierarchy, how can the author avoid it?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java [hunk 5]
```diff
@@ -184,10 +185,20 @@ protected TagletWriter(boolean isFirstSentence) {
     protected abstract Content snippetTagOutput(Element element, SnippetTree snippetTag, StyledText text,
                                                 String id, String lang);
 
+    /**
+     * Returns the output for one or more {@code @spec} tags.
+     *
+     * @param element  the element that owns the doc comment
+     * @param specTags the array of @spec tags.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:192]: Naked tags that aren't meant to be interpreted by javadoc make me uncomfortable.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java [hunk 6]
```diff
@@ -265,11 +265,15 @@ public static Output search(BaseConfiguration configuration, Input input) {
...
+                if (isNonEmpty(output.inlineTags)) {
                     return output;
                 }
             }
         }
         return output;
     }
+
+    private static boolean isNonEmpty(List<? extends DocTree> list) {
+        return list != null && !list.isEmpty();
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java:277]: `output.inlineTags` should never be null. Separately, can this readability change be done in a separate PR? I do have one suitable PR in the works, which can include it.

File: test/langtools/jdk/javadoc/doclet/testConditionalPages/TestConditionalPages.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testConditionalPages/TestConditionalPages.java:2]: Consider adding 6251738 to `@bug`.

File: test/langtools/jdk/javadoc/doclet/testMetadata/TestMetadata.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testMetadata/TestMetadata.java:2]: Same as above.

File: test/langtools/tools/javac/diags/examples/NoTitle.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2012, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline test/langtools/tools/javac/diags/examples/NoTitle.java:2]: It's surprising to see 2012 here.

File: test/langtools/tools/javac/diags/examples/NoURI.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2012, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline test/langtools/tools/javac/diags/examples/NoURI.java:2]: Same as above.

**Review by pavelrappo — COMMENTED** [review]
This is nice work. I haven't seen it in detail, but I'm plannig to do so later.

What's the logistics here? There are two interrelated bugs: 6251738 and 8226279. Are you going to add an 8226279 as an issue to this PR (`/issue add 8226279`)? What about @bug tags in tests? Are they going to mention both issues?

File: src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java [hunk 11]
```diff
@@ -287,6 +287,22 @@ default R visitProvides(ProvidesTree node, P p) {
      */
     R visitSince(SinceTree node, P p);
 
+    /**
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java:290]: The style is consistent with other recent additions that have default methods.  In other words, there is not one but two styles already in play in this file, for the two different kinds of methods.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java [hunk 12]
```diff
@@ -336,6 +337,16 @@ default ReturnTree newReturnTree(boolean isInline, List<? extends DocTree> descr
      */
     SnippetTree newSnippetTree(List<? extends DocTree> attributes, TextTree text);
 
+    /**
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java:340]: I see there is a missing `@since` tag, which I'll fix.
I'd prefer to leave the extra whitespace for now, and maybe cleanup the style of the other comments later.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java [hunk 13]
```diff
@@ -512,8 +512,25 @@ public R visitSnippet(SnippetTree node, P p) {
     /**
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:512]: I'll adjust the order of the tags.

Aside: we should have a guideline somewhere on the recommended order of block tags in a comment,

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java [hunk 14]
```diff
@@ -524,9 +541,7 @@ public R visitStartElement(StartElementTree node, P p) {
     }
 
     /**
-     * {@inheritDoc}
-     *
-     * @implSpec This implementation scans the children in left to right order.
+     * {@inheritDoc} This implementation scans the children in left to right order.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:544]: Bug and/or some variant of cut n paste. I'll ensure to use `@implSpec`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java (hunk 6 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java:277]: I'll back out these lines as long as the tests still pass.

File: test/langtools/jdk/javadoc/doclet/testConditionalPages/TestConditionalPages.java (hunk 7 above)
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testConditionalPages/TestConditionalPages.java:2]: Mmmm, OK

**Comment by jonathan-gibbons** [comment]
> This is nice work. I haven't seen it in detail, but I'm plannig to do so later.
> 
> What's the logistics here? There are two interrelated bugs: 6251738 and 8226279. Are you going to add an 8226279 as an issue to this PR (`/issue add 8226279`)? What about @bug tags in tests? Are they going to mention both issues?

For now, I'll track both, here in the PR and in tests. If this becomes impractical, I'll retain the old/original one and close the newer one.  The old one described the problem; the newer one described a specific solution, although the actual implementation has move on from that proposed in 8226279.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeFactory.java:340]: Added @since. Removed the whitespace for now.

File: src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java [hunk 15]
```diff
@@ -287,6 +287,22 @@ default R visitProvides(ProvidesTree node, P p) {
...
+     * @implSpec Visits the provided {@code SpecTree} node
+     * by calling {@code visitOther(node, p)}.
+     *
+     * @param node the node being visited
+     * @param p a parameter value
+     * @return a result value
+     *
+     * @since 19
+     */
+    default R visitSpec(SpecTree node, P p) {
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java:302]: Nit: this method breaks the alphabetic order of visitor methods. Although `visitOther` also breaks it, we should not address it in this PR.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java [hunk 16]
```diff
@@ -512,8 +512,25 @@ public R visitSnippet(SnippetTree node, P p) {
     /**
      * {@inheritDoc}
      *
+     * @param node {@inheritDoc}
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:515]: There's something wrong with the `@implSpec` tags in this and the immediately following two methods. Might be an artifact of the patch.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java [hunk 17]
```diff
@@ -1525,6 +1525,24 @@ private List<DCTree> tagAttrs() {
                 }
             },
 
+            // @spec url label
+            new TagParser(TagParser.Kind.BLOCK, DCTree.Kind.SPEC) {
+                @Override
+                public DCTree parse(int pos) throws ParseException {
+                    skipWhitespace();
+                    DCText url = inlineWord();
```
Comment (pavelrappo) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java:1533]: Nit: call it `uri` for consistency.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties [hunk 18]
```diff
@@ -406,6 +414,22 @@ doclet.Error_copying_legal_notices=Error while copying legal notices: {0}
...
+    The title "{0}" is used for {1} different external specifications in @spec tags
+
+# 0: name; 1: url
+doclet.extSpec.title.url=\
+    title: "{0}", url: {1}
+
+# 0: url; 1: name
+doclet.extSpec.url.title=\
+    url: {0}, title: "{1}"
+
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:432]: Use `uri` in all 4 resources?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2001, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:2]: 2001?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java [hunk 20]
```diff
@@ -1045,6 +1094,14 @@ public int sourceTabSize() {
         return sourceTabSize;
     }
 
+    /**
+     * Argument for command-line option {@code --spec-base-URI}.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java:1098]: I think the option is lower-cased `--spec-base-uri`.

File: test/langtools/tools/javac/diags/examples/NoTitle.java (hunk 9 above)
Comment (pavelrappo) [inline test/langtools/tools/javac/diags/examples/NoTitle.java:2]: 2012? Is it an old file from the previous incarnation of the `@spec` tag?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java (hunk 6 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/DocFinder.java]: done

File: test/langtools/tools/javac/diags/examples/NoTitle.java (hunk 9 above)
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/diags/examples/NoTitle.java:2]: It was derived from a file created back then. But, it's no great less to lose the 2012.

File: test/langtools/tools/javac/diags/examples/NoURI.java (hunk 10 above)
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/diags/examples/NoURI.java:2]: Same as above

File: src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java (hunk 15 above)
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/doctree/DocTreeVisitor.java:302]: Actually, fixed. Another merge issue.

File: src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java (hunk 16 above)
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/source/util/DocTreeScanner.java:515]: Believed fixed.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java (hunk 17 above)
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java:1533]: Update, the terminology is converging on `URL` in all public places, with `java.net.URI` being used as needed within the doclet implementation.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties (hunk 18 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:432]: Converging on URL.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java (hunk 20 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java:1098]: yes, and now `--spec-base-url`.

File: test/langtools/tools/javac/diags/examples/NoTitle.java (hunk 9 above)
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/diags/examples/NoTitle.java:2]: From the file it was derived from.   It's always a toss-up when cloning and modifying an old file.  2012 removed.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java [hunk 21]
```diff
@@ -295,6 +298,12 @@ public enum ModularityMismatchPolicy {
      */
     private int sourceTabSize;
 
+    /**
+     * Argument for command-line option {@code --spec-base-URI}.
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java:302]: Have my earlier comment on the upper case in `--spec-base-URI` disappeared? I cannot seem to find it.

Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java:302]: Ah, found it: there are two of those cases.

**Review by pavelrappo — COMMENTED** [review]
I appreciate your perseverance in this review, Jon! A note to self: when we have the next round of review, most URIs will have likely turned into URLs.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java (hunk 19 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:2]: removed

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java (hunk 21 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/BaseOptions.java]: I believe these have all been fixed

File: src/jdk.compiler/share/classes/com/sun/source/doctree/DocTree.java [hunk 22]
```diff
@@ -218,6 +218,14 @@ enum Kind {
          */
         SNIPPET("snippet"),
 
+        /**
+         * Used for instances of {@link SpecTree}
+         * representing an {@code @spec} tag.
+         *
+         * @since 19
```
Comment (jddarcy) [inline src/jdk.compiler/share/classes/com/sun/source/doctree/DocTree.java:225]: Please correct the `@since` tag.

File: src/jdk.compiler/share/classes/com/sun/source/util/SimpleDocTreeVisitor.java [hunk 23]
```diff
@@ -463,6 +463,23 @@ public R visitSnippet(SnippetTree node, P p) {
...
+     * {@inheritDoc}
+     *
+     * @implSpec This implementation calls {@code defaultAction}.
+     *
+     * @param node {@inheritDoc}
+     * @param p    {@inheritDoc}
+     *
+     * @return  the result of {@code defaultAction}
+     *
+     * @since 19
```
Comment (jddarcy) [inline src/jdk.compiler/share/classes/com/sun/source/util/SimpleDocTreeVisitor.java:476]: Please update the `@since` tag.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java (hunk 5 above)
Comment (irisclark) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java:203]: `{@code @spec}` similar to line 200?

**Review by hns — COMMENTED** [review]
All looks very good! The only place that looks suspicious to me is the `inherit` implementation in `SpecTaglet.java`, which also seems not to be covered by the test. 

I guess the best way to examine the output of the new tag is via test output. I'll have a look at that to see the HTML and CSS in action before giving a thumbs up.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+    public void inherit(Input input, DocFinder.Output output) {
+        List<? extends SpecTree> tags = input.utils.getSpecTrees(input.element);
+        if (!tags.isEmpty()) {
+            CommentHelper ch =  input.utils.getCommentHelper(input.element);
+            output.holder = input.element;
+            output.holderTag = tags.get(0);
+            output.inlineTags = input.isFirstSentence
+                    ? ch.getFirstSentenceTrees(output.holderTag)
+                    : ch.getReference(output.holderTag);
+        }
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:65]: Thanks for the comment; I'll investigate

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:65]: Good catch;

Yes, the use of getReference is indeed questionable, given its implementation, which is specific to `@see`.   Given ongoing recent work, it's also not clear that the `@spec` tag should be inheritable.

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:65]: For now, I'll change it to `getTags`, pending discussions on whether the tag should be inheritable.

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SpecTaglet.java:65]: With the latest update, toy examples show that tag inherits works as expected, as least as far as parity with `@see`.

**Comment by jonathan-gibbons** [comment]
> I don't know if you want to further discuss inheritability of the new tag, but the code looks good to me. I've also looked at the generated output and it also looks good.

I think of `@spec` as a highly specialized form of `@see` with the significant side effect of generating an extra summary page. So, I'm happy/OK that the support for inheritability is on a par with `@see`. It's not great, since it will only inherit one tag (!!) and you cannot specify which one, but the code is sufficiently similar so when we fix one (later) we should be able to easily fix the other.
